docs/Off-line-Signatures-QR-Code.md

```diff
@@ -192,6 +211,55 @@ Generate the QR code from `offlineData`. Code example in Java:
             return "data:image/png;base64," + BaseEncoding.base64().encode(bytes);
 ```
 
-### 5. Verify ECDSA signature of offlineData
+The user can now scan the QR code via the mobile token app.
+
+The value of `nonce` must be stored somewhere - for example on a browser level, either as a hidden HTML form input field or as a React.js variable, or alternatively, it can be stored in the user session by the specific operation. The `nonce` value is required for later offline signature validation, see step 3.
+
+### 2. Computing Signatures on Mobile Device
+
+#### 2.1. Verify ECDSA Signature of Offline Data
+
+After user scans the QR code using a mobile app, the `ECDSA_SIGNATURE` should be validated on mobile device to verify authenticity of received data by taking contents of `offlineData` before the `ECDSA_SIGNATURE` and computing the ECDSA signature using `KEY_SERVER_PRIVATE`. Both signatures must match before continuing with the offline data signature verification.
+
+#### 2.2. Computing the Signature
+
+Mobile device prompts the user for the PIN code or use of a biometry and computes 4x4 digit long authentication code to be rewritten manually.
+
+### 3. Validating the Signature
+
+#### 3.1. Processing The User Input
+
+After user enters 4x4 digits in the browser, the value must be converted into a standard PowerAuth signature format that uses 2x8 digits. For example, an offline signature `1234-5678-9012-3456` needs to be converted into `12345678-90123456` (removing 1st and 3rd dash).
+
+#### 3.2. Preparing Signature Base String
+
+Now, you need to prepare a normalized data package called "signature base string". This is the payload that mobile app used to compute the signature - you need to have the same signature base string in order to be able to verify the signature.
+
+To compute the signature base string, you need:
+
+- `data` (as obtained in 1.1.)
+- `nonce` value (as obtained in 1.2)
+- two static constants: `POST` and `/operation/authorize/offline`
+
+The [algorithm for signature data normalization](https://developers.wultra.com/docs/develop/powerauth-crypto/Computing-and-Validating-Signatures#normalized-data-for-http-requests) is available in the cryptography description. The Java code for computing the normalized signature base string follows:
+
+```java
+String signatureBaseString
+    = PowerAuthHttpBody.getSignatureBaseString(
+        "POST",
+        "/operation/authorize/offline",
+        BaseEncoding.base64().decode(nonce),
+        data.getBytes()
+);
+```
```

Comment:
We should add a link to the implementing class, it is hard to find:

https://github.com/wultra/powerauth-crypto/blob/master/powerauth-java-http/src/main/java/io/getlime/security/powerauth/http/PowerAuthHttpBody.java

Reply:
OK